Group sorted array

palegre-tiny opened the pull request:
Added a new aggregation function:

Usage 1:
**groupSortedArray**(n)(param1)

It returns an array with the n first values from a field param1 sorted by itself.
```
 select groupSortedArray(2)(number) from numbers(100)
┌─groupSortedArray(2)(number)─┐
│ [0,1]                       │
└─────────────────────────────┘
```

Usage 2:
**groupSortedArray**(n)(param1, param2)

It returns an array with the n first values from a field param1 sorted param2 (field or expression).

SELECT groupSortedArray(5)(number, number) FROM numbers(1000000)

```
SELECT groupSortedArray(5)(number, -number) FROM numbers(100)
┌─groupSortedArray(5)(number, negate(number))─┐
│ [99,98,97,96,95]                            │
└─────────────────────────────────────────────┘
```

palegre-tiny committed:
Add groupSortedArray() function

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+
+    void add(AggregateDataPtr __restrict, const IColumn **, size_t, Arena *) const override { }
+
+    void addBatchSinglePlace(
```
Using addBatchSinglePlace() requires to calculate different agregations and combine all of them in the final result. I am also adding other optimizations like reading reading the first values (if weighted) only for the final results of each aggregation. It saves a lot of CPU when reading strings.

Algunenano replied:
I'm confused. What's the real benefit of implementing `addBatchSinglePlace`?
Having `addBatchSinglePlace` opens up the posibility of working with a full batch in interesting ways, like finding the batch top N elements and keeping only those (with a partial sort) and things like that, but if you are iterating over all elements and calling `add` then you could use the default implementation and avoid complexity, right?

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+#include <IO/VarInt.h>
+#include <IO/WriteBuffer.h>
+#include <IO/WriteHelpers.h>
+
```
This file contains the code for the internal data set I am using.
It may use single value implemented using multiset or a pair (value and weigth) implemented using multimap.
It also support different value types so, to just provide one final class I am doing some complex staff with templates and inheritance here.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+namespace DB
+{
+template <typename Storage>
+struct AggregateFunctionGroupSortedArrayDataBase
```
This is just a base class for common code for both std::multimap and std::set operations.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+    AggregateFunctionGroupSortedArrayDataBase(UInt64 threshold_ = DEFAULT_THRESHOLD) : threshold(threshold_) { }
+
+    virtual ~AggregateFunctionGroupSortedArrayDataBase() { }
+    inline void narrowDown()
```
Items are ordered so we never allow more items than requested (defined by 'threshold' variable)

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+template <typename Storage>
+struct AggregateFunctionGroupSortedArrayDataBase
+{
+    typedef typename Storage::value_type ValueType;
```
This is the key, 'typename Storage' defines what kind of std storage we are using, so we define here a std::multiset<X>::value_type or std::multimap<X,Y>::value_type, so here we operate with this abstraction.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+};
+
+template <typename T>
+struct AggregateFunctionGroupSortedArrayData<T, true> : public AggregateFunctionGroupSortedArrayDataBase<std::multimap<Int64, T>>
```
for the multimap implementation (weighted) we always use an integer value for the weight.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+
+    void add(T item, Int64 weight)
+    {
+        if (weight <= last || count < Base::threshold)
```
We keep last stored value (last item.second from values) and count (values.size()) to save some CPU.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+};
+
+template <>
+struct AggregateFunctionGroupSortedArrayData<StringRef, false> : public AggregateFunctionGroupSortedArrayDataMSet<StringRef>
```
Template speficication when type is StringRef, in this case we don't pre-evaluate value with the last element and we just let the set decide the order.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+};
+
+template <typename T>
+struct AggregateFunctionGroupSortedArrayData<T, false> : public AggregateFunctionGroupSortedArrayDataMSet<T>
```
And for the rest of cases we optimize not playing with the map if the elements is not minor than the last value. Here the declaration for 'last' uses the same type 'T'.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+};
+}
+
+#undef DEFAULT_THRESHOLD
```
Please make it a constant somewhere and not a macro

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+    {
+        if (item <= last || count < Base::threshold)
+        {
+            ++ count;
```
2 things:
* count is never decresased, even when adding extra values. Is that ok?
* Style looks odd.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+            ++ count;
+            Base::values.insert(item);
+            Base::narrowDown();
+            last = *(--Base::values.end());
```
Why do you need to keep last?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeVarT(UInt64(values.size()), buf);
```
`writeVarUInt` for clarity?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,174 @@
+    if constexpr (std::is_same_v<T, StringRef>)
+        writeBinary(item, buf);
+    else
+        writeVarUInt(item, buf);
```
Does this work when T isn't a 64 size type?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,137 @@
+    AggregateFunctionPtr createAggregateFunctionGroupSortedArray(
+        const std::string & name, const DataTypes & argument_types, const Array & params, const Settings *)
+    {
+        if (!is_weighted)
```
constexpr?

palegre-tiny replied:
The topK() code where I took it from didn't use it, anyways I can add it :+1:  .

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,137 @@
+#include <Common/FieldVisitorConvertToNumber.h>
+
+
+static inline constexpr UInt64 TOP_K_MAX_SIZE = 0xFFFFFF;
```
Name doesn't really match. There might be a constant somewhere with the largest supported array internally (which isn't always respected), that might be a better alternative.

palegre-tiny replied:
Well, I took it from topk() and I forgot to change the name.
I see many different declaration for Array size limit, for similar purposes they declare it to MAX 24 bit integer as well, I don't find a generic one.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,137 @@
+        }
+
+        UInt64 threshold = 10; /// default values
+        UInt64 load_factor = 3;
```
What is load_factor? I don't see it being used anywhere

palegre-tiny replied:
I will remove it, again I took it from topk() but I dont see any sense here.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,137 @@
+
+        if (!res)
+            throw Exception(
+                "Illegal type " + argument_types[0]->getName() + " of argument for aggregate function 4 " + name,
```
Extra " 4 "?

palegre-tiny replied:
yes, by mistake, removing....

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+
+namespace DB
+{
+template <typename TT, bool is_plain_column>
```
Why `TT` and not `T`?

palegre-tiny replied:
I tried to add this function inside the class so I chose a diferent name, but finally it is here and I can use 'T' as normal.

Algunenano replied:
Let's go back to T if it's the only type

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+    }
+    else
+    {
+        return column->getUInt(row);
```
Why is this reading a UInt and not a type based on the column type? Maybe I'm missunderstanding the function but I understand that is something like:
FUNC(LIMIT)(COLUMN1, [COLUMN2])
that returns an array topN elements of Column1, optionally ordered by Column2 and otherwise ordered by themselves (if they are comparable)
-> LIMIT: Numbers of element to keep
-> COLUMN1: Values to keep. This can be of any type (at least any type that can be held in an array, which I guess is everything)
-> COLUMN2: Order of each element. This can only be of type Int64.

palegre-tiny replied:
I took it from topk(), they only use UInt64 for second parameter, I will review if I can use with any numeric value here.

palegre-tiny replied:
Your interpretation is correct, I have some incoherences with this UInt64/Int64.
The problem here, as far as I can see, is the comparison with negative values, apparently topk() doesn't support it but I understand we can get lowest values using **group...(a, b)** or highest using **group....(a,-b)** and then it requires supporting signed. But at the same time I am always storing UInt64, this is not correct.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+        this->data(place).threshold = threshold;
+    }
+
+    String getName() const override { return is_weighted ? "groupSortedArrayWeighted" : "groupSortedArray"; }
```
constexpr

palegre-tiny replied:
I understand this is a template and the compilator will translate to **return (TRUE):'A':'B'** so my guess it will be compiled as **return 'A'**.
Even more, as far as I know using constexpt with ternary operator is not allowed, so I have to use if constexpr...

Algunenano replied:
constexpr avoids the need to guess what the compiler will do. You don't need a ternary operator here at all

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+            return false;
+    }
+
+    void add(AggregateDataPtr __restrict, const IColumn **, size_t, Arena *) const override { }
```
What is going on here? Why isn't this implemented?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+
+        State & data = this->data(place);
```
Why is this mutex needed? Is it possible to have `add` or `addBatch...` to the same object in parallel? I don't recall seeing anything like this anywhere else.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+        {
+            StringRef str = column->getDataAt(row);
+            auto ptr = arena->alloc(str.size);
+            std::copy(str.data, str.data + str.size, ptr);
```
This doesn't seem very efficient if you don't end up using the value. Also, when is this memory freed? Is it going to keep all the strings allocated until the end of the aggregation even if they aren't used?

palegre-tiny replied:
Yes, maybe I broke this efficiency with the last optimization. I will review.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,182 @@
+            if (if_argument_pos >= 0)
+            {
+                const auto & flags = assert_cast<const ColumnUInt8 &>(*columns[if_argument_pos]).getData();
+                for (size_t i = 0; i < batch_size; ++i)
```
Style. I personally would never avoid `{}` if the "content" of the for is larger than a single line.

Algunenano replied:
Please add brackets to anything that goes beyond a single line.

Algunenano commented on tests/queries/0_stateless/02158_groupsortedarray.sql:
```diff
@@ -0,0 +1,20 @@
+
+SELECT groupSortedArrayWeighted(10)(number) from numbers(100); -- { serverError 42 }
+
+SELECT groupSortedArrayWeighted(5)(number, text) FROM (select toString(number) as text, number from numbers(10)); -- { serverError 43 }
```
Consider testing with other types aside from String and Int64. I'd recommend testing things like `Float64`, `Nullable(Float32)`, `Int256`, etc. This is because I fear that the serialization / deserialization isn't correct, so testing with -State and -Merge should bring some peace of mind.

Also, since you have reimplemented addBatch, which covers the -If suffix aggregation, I would test that too.

Lastly, since this is intended to use as a better performance topK (keeping duplicates too) then a performance check is in order to measure things out.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,127 @@
+        const std::string & name, const DataTypes & argument_types, const Array & params, const Settings *)
+    {
+        if (!is_weighted)
+        {
```
Let's add the constexpr for all the things that are

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.cpp:
```diff
@@ -0,0 +1,127 @@
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+
+        UInt64 threshold = 10; /// default values
```
Maybe move this with the other constant `GROUP_SORTED_ARRAY_MAX_SIZE`.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+#include <DataTypes/DataTypeArray.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <base/logger_useful.h>
```
Extra unnecessary include left from dev

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <base/logger_useful.h>
+#include "AggregateFunctionGroupSortedArrayData.h"
```
Most of all the includes are done using `<` so I would use that too here.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+        return column->getUInt(row);
+    }
+}
+/*
```
This old code should be removed if it's no longer used.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+*/
+
+template <typename T>
+void getFirstNElements(const T * data, int num_elements, int threshold, uint64_t * results)
```
Likely none of these types or the ones above should be `int`. Please use the same types as the input of the aggregations. Also`uint64_t` -> `UInt64`.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+    int z;
+    for (int i = 0; i < num_elements; i++)
+    {
+        //We starting from the highest values and we look for the immediately lower than the given one
```
`We are starting`

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
+                results[z + 1] = results[z];
+            if (++current_max > threshold)
```
Wouldn't it be simpler, and avoid this check, if you prepopulate the results array with the first `threshold` elements, sort them, and then continue knowing that current_max is always == threshold?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+
+    UInt64 threshold;
+    DataTypePtr & input_data_type;
+    mutable std::mutex mutex;
```
Unused

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+        if constexpr (is_weighted)
+        {
+            StringRef ref = columns[1]->getRawData();
+            UInt64 values[batch_size];
```
Why are you copying the values instead of using the const data directly. It seems to me you don't need to copy anything right?

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+            StringRef ref = columns[1]->getRawData();
+            UInt64 values[batch_size];
+            memcpy(values, ref.data, batch_size * sizeof(UInt64));
+            UInt64 * bestRows = new UInt64[this->threshold];
```
Please, no new and no delete[]. No-no-no-no.no-no :D Either use a VLA like you did with values or put it inside a smart pointer if you want it to go to the heap.

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+
+    void addBatchSinglePlace(
+        size_t batch_size, AggregateDataPtr place, const IColumn ** columns, Arena * arena, ssize_t if_argument_pos) const override
+    {
```
The function is ignoring the if parameter and thus is broken everywhere:

```
SELECT groupSortedArrayWeightedIf(10)(a, -b, b > 90) AS state
FROM
(
    SELECT
        number AS a,
        number AS b
    FROM numbers(100)
)

Query id: b1dcb635-1763-4f58-98be-bd6e6ebd511c

┌─state───────────────┐
│ [8,7,6,5,4,3,2,1,0] │
└─────────────────────┘

1 rows in set. Elapsed: 0.001 sec. 
```

```
SELECT topKWeightedIf(10)(a, -b, b > 90) AS state
FROM
(
    SELECT
        number AS a,
        number AS b
    FROM numbers(100)
)

Query id: 562d34f5-bdf9-4572-81be-b9e2801b628a

┌─state────────────────────────┐
│ [91,92,93,94,95,96,97,98,99] │
└──────────────────────────────┘

1 rows in set. Elapsed: 0.001 sec. 
```

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+            size_t num_results;
+
+            //First store the first n elements with the column number
+            if (if_argument_pos >= 0)
```
This looks extremely inefficient and not an improvement over topK (in fact because it's bugged for the -If it's worse).

The way I see if you need a function that given an array of UInt64 gives you back the **indexes** of top `threshold` elements and then you access those only.

palegre-tiny replied:
I'm not sure about when 'if_argument_pos' is greater than 0 and the we run this code. But I understand we it is a filter to apply to the rows, we need first and array with all the data to evaluate so honestly speaking I don't know how to do it better.
I took this code from the base class IAggregateFunction::addBatchSinglePlace().

Algunenano commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,248 @@
+        ColumnArray::Offsets & offsets_to = arr_to.getOffsets();
+
+        auto & values = this->data(place).values;
+        int old_size = offsets_to.back();
```
Please no `int`. `int` in general is a bad option for sizes since you don't know how big they are and if they are able to fit the size type. Either use the exact same size returned by back() or simply use `auto`.

palegre-tiny replied:
Yes, I don't need to declare 'old_size' as a variable anymore because we only use it once.

Algunenano commented on tests/queries/0_stateless/02158_groupsortedarray.reference:
```diff
@@ -0,0 +1,12 @@
+[0,1,2,3,4]
```
Seems like the expected order is different from topK:

```
Mordor :) SELECT groupSortedArrayWeighted(number, number) from numbers(100);

SELECT groupSortedArrayWeighted(number, number)
FROM numbers(100)

Query id: cf18af29-2f70-4400-9aff-fbf04530e8ac

┌─groupSortedArrayWeighted(number, number)─┐
│ [0,1,2,3,4,5,6,7,8,9]                    │
└──────────────────────────────────────────┘

1 rows in set. Elapsed: 0.011 sec. 

Mordor :) SELECT topKWeighted(number, number) from numbers(100);

SELECT topKWeighted(number, number)
FROM numbers(100)

Query id: 2ee96f56-9478-457b-9ecb-71ee0d9a3ce8

┌─topKWeighted(number, number)────┐
│ [99,98,97,96,95,94,93,92,91,90] │
└─────────────────────────────────┘

1 rows in set. Elapsed: 0.009 sec. 
```

I'd expect to be exactly the same order.

palegre-tiny replied:
Now that we realised 'weight' doesn't mean 'orderby' we may change the first to use a more appropiate (orderby/ sort /any).
But I would say it is ok to use this order (ascend).

Algunenano commented on tests/queries/0_stateless/02158_groupsortedarray.reference:
```diff
@@ -0,0 +1,12 @@
+[0,1,2,3,4]
+[0,1,2,3,4,5,6,7,8,9]
+[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99]
+[999,998,997,996,995,994,993,992,991,990,989,988,987,986,985,984,983,982,981,980,979,978,977,976,975,974,973,972,971,970,969,968,967,966,965,964,963,962,961,960,959,958,957,956,955,954,953,952,951,950,949,948,947,946,945,944,943,942,941,940,939,938,937,936,935,934,933,932,931,930,929,928,927,926,925,924,923,922,921,920,919,918,917,916,915,914,913,912,911,910,909,908,907,906,905,904,903,902,901,0]
```
This output is clearly wrong, no way 0 follows 901:

```
SELECT groupSortedArrayWeighted(5)(number, -number)
FROM numbers(1000)

Query id: f2c986a7-c5c4-4296-a39e-928dd5ba3f26

┌─groupSortedArrayWeighted(5)(number, negate(number))─┐
│ [999,998,997,996,0]                                 │
└─────────────────────────────────────────────────────┘
```

palegre-tiny replied:
This is happening because we always used UInt64 for weight comparison, I have added another template argument to use signed/unsiged depending on the column type.

Algunenano reviewed (commented):
Aside from the comments in the code itself here are a few things that are missing:
- Docs. Specially important to mention what are the differences with topK in behaviour.
- Tests with complex types (like tuples).
- Performance tests. The performance with strings, and thus tuples and so on, seems it could be improved.

palegre-tiny committed:
fixup! Add groupSortedArray() function

- Fix memory access
- Support any type as sorting parameter
- Fix tests
- Rewrite/simplify function addBatchSinglePlace

palegre-tiny committed:
fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! Add groupSortedArray() function

fixup! Add groupSortedArray() function

fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! Evolve to TinySum

palegre-tiny committed:
fixup! Evolve to TinySum

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

palegre-tiny committed:
fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! fixup! Add groupSortedArray() function

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

mergify[bot] committed:
Merge branch 'master' into groupSortedArray
